### PR TITLE
Build predicate by type

### PR DESCRIPTION
**Pull request by MollardMichael**

### CODE REVIEW

File: src/main/kotlin/com/sipios/springsearch/CriteriaParser.kt [hunk 1]
```diff
@@ -24,7 +24,7 @@ class CriteriaParser<T> {
         return visitor.visit(parser.input())
     }
 
-    private fun getParser(queryString: String) : QueryParser {
+    private fun getParser(queryString: String): QueryParser {
```
Comment (wrousseau) [inline src/main/kotlin/com/sipios/springsearch/CriteriaParser.kt]: We don't have a linter that detected that?

File: src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt [hunk 2]
```diff
@@ -17,24 +19,24 @@ import java.util.ArrayList
...
-            SearchOperation.STARTS_WITH -> return builder.like(nestedRoot.get(criteriaKey), criteria.value + "%")
-            SearchOperation.ENDS_WITH -> return builder.like(nestedRoot.get(criteriaKey), "%" + criteria.value)
-            SearchOperation.CONTAINS -> return builder.like(builder.lower(nestedRoot.get<String>(criteriaKey).`as`(String::class.java)), "%" + criteria.value + "%")
-            SearchOperation.DOESNT_START_WITH -> return builder.notLike(nestedRoot.get(criteriaKey), criteria.value + "%")
-            SearchOperation.DOESNT_END_WITH -> return builder.notLike(nestedRoot.get(criteriaKey), "%" + criteria.value)
-            SearchOperation.DOESNT_CONTAIN -> return builder.notLike(builder.lower(nestedRoot.get<String>(criteriaKey).`as`(String::class.java)), "%" + criteria.value + "%")
-            else -> return null
+        val fieldClass = nestedRoot.get<Any>(criteriaKey).javaType.kotlin
+        val strategy = ParsingStrategy.getStrategy(fieldClass)
+        val value: Any?
```
Comment (wrousseau) [inline src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt]: Is it possible to create a type to specify handled types?

File: src/main/kotlin/com/sipios/springsearch/strategies/BooleanStrategy.kt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.sipios.springsearch.strategies
+
+import kotlin.reflect.KClass
+
+class BooleanStrategy : ParsingStrategy {
+    override fun parse(value: String?, fieldClass: KClass<out Any>): Any? {
```
Comment (wrousseau) [inline src/main/kotlin/com/sipios/springsearch/strategies/BooleanStrategy.kt]: same here for the Any

File: src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+class DateStrategy : ParsingStrategy {
+    private val sdf: DateFormat = StdDateFormat()
+
+    override fun buildPredicate(
+        builder: CriteriaBuilder,
+        path: Path<*>,
+        fieldName: String,
+        ops: SearchOperation?,
+        value: Any?
```
Comment (wrousseau) [inline src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt]: could you document that function?

File: src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    ): Predicate? {
+        return when (ops) {
+            SearchOperation.GREATER_THAN -> builder.greaterThan(path.get<Date>(fieldName), value as Date)
+            SearchOperation.LESS_THAN -> builder.lessThan(path.get<Date>(fieldName), value as Date)
+            else -> super.buildPredicate(builder, path, fieldName, ops, value)
+        }
+    }
+
+    override fun parse(value: String?, fieldClass: KClass<out Any>): Any? {
+        return sdf.parse(value)
```
Comment (wrousseau) [inline src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt]: same here

**Comment by FrankHeijden** [comment]
Currently it's not yet possible to search case insensitive here, fixed that in #15

**Comment by noxsicarius** [comment]
What is the general progress on getting this item added? I see this was completed a year ago, but has not been added.

File: src/main/kotlin/com/sipios/springsearch/CriteriaParser.kt (hunk 1 above)
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/CriteriaParser.kt:27]: We do now

File: src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import com.sipios.springsearch.SearchOperation
+import java.text.DateFormat
+import java.util.Date
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.Path
+import javax.persistence.criteria.Predicate
+import kotlin.reflect.KClass
+
+class DateStrategy : ParsingStrategy {
+    private val sdf: DateFormat = StdDateFormat()
```
[the remark this answers is not in the document]
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt:13]: I renamed it

File: src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt (hunk 2 above)
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt:29]: I did not find a solution for this problem. 
Kotlin seems to currently have no solution for this : https://discuss.kotlinlang.org/t/union-types/77/67

File: src/main/kotlin/com/sipios/springsearch/strategies/BooleanStrategy.kt (hunk 3 above)
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/strategies/BooleanStrategy.kt:6]: See above

File: src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt (hunk 5 above)
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt]: Done

File: src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt (hunk 4 above)
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/strategies/DateStrategy.kt:20]: Yes

File: src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt [hunk 7]
```diff
@@ -1,40 +1,42 @@
...
         val criteriaKey = nestedKey[nestedKey.size - 1]
-
-        when (criteria.operation) {
-            SearchOperation.EQUALS -> return builder.equal(nestedRoot.get<String>(criteriaKey), criteria.value)
-            SearchOperation.NOT_EQUALS -> return builder.notEqual(nestedRoot.get<String>(criteriaKey), criteria.value)
-            SearchOperation.GREATER_THAN -> return builder.greaterThan(nestedRoot.get<Double>(criteriaKey), criteria.value!!.toDouble())
-            SearchOperation.LESS_THAN -> return builder.lessThan(nestedRoot.get<Double>(criteriaKey), criteria.value!!.toDouble())
-            SearchOperation.STARTS_WITH -> return builder.like(nestedRoot.get(criteriaKey), criteria.value + "%")
-            SearchOperation.ENDS_WITH -> return builder.like(nestedRoot.get(criteriaKey), "%" + criteria.value)
-            SearchOperation.CONTAINS -> return builder.like(builder.lower(nestedRoot.get<String>(criteriaKey).`as`(String::class.java)), "%" + criteria.value + "%")
```
Comment (FrankHeijden) [inline src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt:32]: So it seems the contains operator was comparing in ignored case in the old method; however with the changes proposed here, this functionality got removed. My PR #10 was quick solution to fix the problem, however with the recent changes it is outdated. Is the intention still to keep the ignored casing here -- or move away from this?

File: src/main/kotlin/com/sipios/springsearch/strategies/ParsingStrategy.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        fieldName: String,
+        ops: SearchOperation?,
+        value: Any?
+    ): Predicate? {
+        return when (ops) {
+            SearchOperation.EQUALS -> builder.equal(path.get<Any>(fieldName), value)
+            SearchOperation.NOT_EQUALS -> builder.notEqual(path.get<Any>(fieldName), value)
+            SearchOperation.STARTS_WITH -> builder.like(path.get(fieldName), "$value%")
+            SearchOperation.ENDS_WITH -> builder.like(path.get(fieldName), "%$value")
+            SearchOperation.CONTAINS -> builder.like((path.get<String>(fieldName).`as`(String::class.java)), "%$value%")
```
Comment (FrankHeijden) [inline src/main/kotlin/com/sipios/springsearch/strategies/ParsingStrategy.kt:44]: (In addition to my other comment) The `.lower` got removed in this specific line, and in the SearchOperation `DOESNT_CONTAIN`.

File: src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt (hunk 7 above)
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/SpecificationImpl.kt:32]: Indeed, we want to have case sensitive lookups by default. However, we want to implement a parameter to the annotation to permit case insensitive lookups for cases like yours. The goal is to be able to set case sensitivity by fields independently.

File: src/main/kotlin/com/sipios/springsearch/strategies/ParsingStrategy.kt (hunk 8 above)
Comment (luc-boussant) [reply src/main/kotlin/com/sipios/springsearch/strategies/ParsingStrategy.kt:44]: See response above :)

**Pull request merged**